src/fn/watch-project.js

```diff
@@ -0,0 +1,206 @@
+    }
+    return false;
+};
+
+const processContactForm = (fileName) => {
+    if (convertForms.FORM_FILE_MATCHER(fileName)) {
+        eventQueue.enqueue(async () => {
+            await convertContactForm([fileName.split('.')[0]]);
+            return fileName;
+        });
+        return true;
+    }
+    if (uploadForms.FORM_FILE_MATCHER(fileName)) {
+        eventQueue.enqueue(async () => {
+            await uploadContactForms([fileName.split('.')[0]]);
+            return fileName;
+        });
+        return true;
+    }
+    return false;
+};
+
+const processConfigFiles = (api, fileName) => {
+    if (fileName === 'app_settings.json') {
+        eventQueue.enqueue(async () => {
+            await uploadAppSettings(api);
+            return fileName;
+        });
+        return true;
+    }
+    if (fileName.match(/.+\.js$/) || fileName.match(/^[\w]+\.json$/)) {
+        eventQueue.enqueue(async () => {
+            await compileAppSettings();
+            return fileName;
+        });
+        return true;
+    }
+    return false;
+};
+
+const watchProject = {
+    watch: async (api, blockFn, callback = {}, uploadStateOnStart = false) => {
+        if (uploadStateOnStart) {
+            await uploadInitialState(api);
+            info('Initial State Uploaded');
+        }
+
+        eventQueue = new Queue({ concurrent: 1, start: true });
+
+        fsEventsSubscription = await watcher.subscribe(environment.pathToProject, async (err, events) => {
+            if (err) {
+                error(err);
+                return;
+            }
+            for (const event of events) {
+                if (event.type !== 'update' && event.type !== 'create') {
+                    continue;
+                }
+                const changePath = event.path;
+                const parsedPath = path.parse(changePath);
+
+                if (parsedPath.dir.replace(environment.pathToProject, '').startsWith(`/${APP_FORMS_PATH}`)) {
+                    const relativeFormsDir = parsedPath.dir.replace(environment.pathToProject, '');
+                    const fileName = parsedPath.base;
+                    if (!processAppForm(fileName) && !relativeFormsDir.endsWith(APP_FORMS_PATH)) {
+                        const formMediaDir = relativeFormsDir.replace(relativeFormsDir, '');
+                        processAppFormMedia(formMediaDir, fileName);
+                    }
+                    continue;
+                }
+
+                if (parsedPath.dir === path.join(environment.pathToProject, CONTACT_FORMS_PATH)) {
+                    const fileName = parsedPath.base;
+                    processContactForm(fileName);
+                    continue;
+                }
+
+                if (parsedPath.dir === path.join(environment.pathToProject, TRANSLATIONS_DIR_PATH)) {
+                    await uploadCustomTranslations();
+                    const fileName = parsedPath.base;
+                    if (callback) callback(fileName);
```

Comment:
It's unclear why the callback is only called for some actions, and not called for others. 
Best practice is to design your tests so that they work with your code, and not the other way around. If this is just required for a test, can the test be refactored so it does not require this or at least be consistent with the callbacks?

Reply:
It's called for all functions, the actions added to the event queue return the filename when resolved https://github.com/medic/cht-conf/blob/b715eea8a62176c82f49c25c6858b05f9ccc4716/src/fn/watch-project.js#L185-L189

Reply:
I'm not sure if it's possible to run the tests without a way of getting notified of changes but let me see if i can figure out a way

Reply:
One option would be to watch Couchdb changes and run your assertions when detecting changes. But that't not necessarily achievable, since some actions make many changes that you'd need to keep track of...